# Understand EPSG, WKT, and Other CRS Definition Styles

### CRS formats

THere are numerous formats that are used a document a CRS. 3 common formats include:
1. proj.4
2. EPSG
3. Wel-Known Text (WKT) formats

Often you have CRS information in 1 format and you need to translate that CRS into a different format to use in a tool like Python.

1 of the msot powerful websites to look at CRS strings is Spatialreference.org. You can search on the site to find an EPSG code. 

### PROJ or PROJ.4 strings

PROJ.4 strings are a compact way to identify a spatial or CRS PROJ.4 strings are 1 of the formats that Geopandas can accept. However, note that many libraries are moving towards the more concise EPSG format.

Using the PROJ.4 syntax, you specify the complete set of parameters including the ellipse, datum, projection units, and projection definition taht define a particular CRS.

#### Breakdown the PROJ.4 format

Below is an example of a proj.4 string:

+proj=utm+zone=11 +datum=WGS84 +units=m +no_defs +ellps=WGS84 +towgs84=0,0,0

Notice that the CRS info is structured using a string of characters and numers that are combined using + signs. The CRS for your data are in the proj4 format. The string contains all of the individual CRS elements that Python or another GIS might need. Each element is specified with a + sign, similar to how a .csv file is delimited or broken up by a , ..

You can breakdown the proj4 string into its individual components (again, separated by + signs) as follows:
- +proj=utm: the projection is UTM, UTM has several zones.
- +zone=11: the zone is 11 which is a zone on the west coast, USA.
- datum=WGS84: the datum WGS84 (the datum refers to the 0.0 reference for the coordinate system used in the projection)
- +units=m: the units for the coordinates are in METERS.
- +ellps=WGS84: the ellipsoid (how the earth's roundness is calculated for the data is WGS84

Note that the zone is unique to the UTM projection. Not all CRS will have a zone.

### Geographic (lat/long) Proj.4 String

Look at another CRS definition.

+proj=longlat +datum=WGS84 +no_defs +ellips=WGS84 +towgs84=0,0,0

This is a lat/long or geographic projection. The components of the proj4 string are broken down below.
- +proj=longlat: the data are in a geographic (lat and long) coordinate system.
- datum=WGS84: the datum WGS84 (the datum refers to the 0,0 reference for the coordinate system used in the projection)
- +ellps=WGS84: the ellipsoid (how the earth's roundness is calculated) is WGS84

Note that there are no specified units above. This is because the geographic CRS is in lat and long which is most often recorded in Decimal Degrees.

* Data Tip: THe last portion of each prj4 string is *towgs84=0,0,0 . This is a conversion factor that is used if a datum conversion is required.

### EPSG codes

The EPSG codes area 4-5 digit numbers that represent CRSs definitions. The acronym EPSG comes from the, now defunt. European Petroleum Survey Group. Each code is a 4-5 digit number which represents a particular CRS definition.

In [9]:
import os
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import geopandas as gpd
from shapely.geometry import Point
import earthpy as et

# Set work dir & get data
data = et.data.get_data("spatial-vector-lidar")
os.chdir(os.path.join(et.io.HOME, "earth-analytics"))

# Import world boundary shapefile
worldBound_path = os.path.join("data", "spatial-vector-lidar", "global",
                              "ne_110m_land", "ne_110m_land.shp")

worldBound = gpd.read_file(worldBound_path)

worldBound.crs

{'init': 'epsg:4326'}

NOtice that the CRS returned, consists of 2 parts:
1. "init" which tells python that a CRS definition (i.e. EPSG code) will be provided and
2. the EPSG code itself epsg:4326

### How to create a CRS object

You often need to define the CRS for a spatial object. For example in the previous lessons, you created new spatial point layers, and had to define the CRS that the point x, y locations were in:

1. You manually created an array for a single point:(x, y).
2. You turned that x, y point into a shapely point object.
3. FInally convert that point object to a pandas GeoDataFrame

In [10]:
# Create a numpy array with x, y location of Boulder
boulder_xy = np.array([[476911.31, 4429455.35]])

# Create shapely point object
boulder_xy_pt = [Point(xy) for xy in boulder_xy]

# Convert to spatial dataframe - geodataframe -- assign the CRS using epsg code
boulder_loc = gpd.GeoDataFrame(boulder_xy_pt,
                              columns=["geometry"],
                              crs={"init": "epsg:2957"})

# View crs of new spatial points object
boulder_loc.crs

{'init': 'epsg:2957'}

#### WRT or Well-Known Text

it's useful to recognise this format - ArcMap and ENVI use this format. Well-Known Text (WKT) is a compacted machine- and human-readable representation of geometric objects. It defines elements of CRS definitions using a combination of brackets [] and elements separated by commas (,).

Here is an example of WKT for WGS84 geographic:

GEOGCS["GCS_WGS_1984",DATUM["D_WGS_1984",SPHEROID["WGS_1984",6378137,298.257223563]],PRIMEM["Greenwich",0],UNIT["Degree",0.017453292519943295]]

Notice here that the elements are described explicitly using caps - for example:
- UNIT
- DATUM

## How to look up a CRS

The most powerful website to look-up CRS info is the spatialreference.org website. This website has a useful search function that allows you to search for strings such as:
- UTM 11N or
- WGS84